### PR TITLE
Feature file upload service

### LINKED ISSUE #26: Original file upload service
### Descriptive summary

Given a configurations in the initializer, upload the original binary file(s) the proper path used by the application (browse everything). There should be the capability to upload the file to a file system path, or to an S3 bucket.

### Expected behavior

The engine initializer should have configurations (values coming from ENV) for a file system path, S3 credentials, and S3 bucket. The service would inspect the configurations to decide where the appropriate storage location is, store a file in that location, and return details about the file and it's current location.

### PR DESCRIPTION
**Pull request by luisgreg99**
fixes #26 

- [x] explore `ActiveStorage` as the framework to handle file uploads
- [x] explore `BrowseEverything` features to handle file uploads
- [x] implement `Hyrax::Migrator::FileUploadService.upload_file_content`

Notes:
- It looks like the most stable version of `ActiveStorage` is only available on Rails 5.2.0: https://edgeguides.rubyonrails.org/active_storage_overview.html#setup
- The recommended rails version for hyrax is 5.1 according to the readme: https://github.com/samvera/hyrax#rails
- Current issue in OD2 to update Rails 5.2: https://github.com/OregonDigital/OD2/issues/314
- Related notes on ActiveStorage: https://bibwild.wordpress.com/2018/10/03/some-notes-on-whats-going-on-in-activestorage/

### CODE REVIEW

File: app/services/hyrax/migrator/services/file_upload_service.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      end
+
+      def upload_file_content
+        if @upload_storage_service == :aws_s3
+          upload_to_s3
+        elsif @upload_storage_service == :file_system
+          upload_to_file_system
+        end
+      end
+
```
Comment (revgum) [inline app/services/hyrax/migrator/services/file_upload_service.rb]: It seems like all of the following methods could be `private`

File: app/services/hyrax/migrator/services/file_upload_service.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      def upload_to_file_system
+        name = File.basename(content_file)
+        dest_name = File.join(file_system_path, name)
+        IO.copy_stream(content_file, dest_name)
```
Comment (revgum) [inline app/services/hyrax/migrator/services/file_upload_service.rb]: Could need a `rescue` block here to log/handle if `IO.copy_stream` fails and why?

File: app/services/hyrax/migrator/services/file_upload_service.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      end
+
+      def upload_to_s3
+        # Get just the file name
+        name = File.basename(content_file)
+        # Create the object to upload
+        obj = aws_s3_resource.bucket(@aws_s3_bucket).object(name)
+        # Upload it      
+        return nil unless obj.upload_file(content_file)
+        # if upload is successful, return presigned url available for 1 hour
```
Comment (revgum) [inline app/services/hyrax/migrator/services/file_upload_service.rb]: I think we should consider surfacing the `aws_s3_url_availability` configuration and default it to something very long.. these urls need to be available in the context of a background job that might not run soon. I'm thinking 24 hours as a default would be a good start, but likely a week plus. :)

File: app/services/hyrax/migrator/services/file_upload_service.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        obj.presigned_url(:get, expires_in: AWS_S3_PRESIGNED_GET_URL_VALID)
+      end
```
Comment (revgum) [inline app/services/hyrax/migrator/services/file_upload_service.rb]: Also probably a good idea for a `rescue` block to handle any exceptions that might be raised by the S3 sdk?

**Comment by revgum** [comment]
Also, can you add to `spec/dummy/config/initializers/hyrax_migrator.rb` with the config defaults and documentation about them?

File: app/services/hyrax/migrator/services/file_upload_service.rb (hunk 4 above)
Comment (luisgreg99) [reply app/services/hyrax/migrator/services/file_upload_service.rb]: I've added a rescue block. I think that's a good idea.

File: app/services/hyrax/migrator/services/file_upload_service.rb (hunk 2 above)
Comment (luisgreg99) [reply app/services/hyrax/migrator/services/file_upload_service.rb]: done

**Pull request merged**